File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_09.py
#     *
#    ***
#   *****
#  *******
# *********
#  *******
#   *****
#    ***
#     *
#----------------------------------
# BRUTE FORCE APPROACH
#----------------------------------
n = 5
# Upper half
for i in range(1, n + 1):
    line = ""
    for j in range(n - i):  # Spaces
        line += " "
    for k in range(2 * i - 1):  # Stars
        line += "*"
    print(line)

# Lower half
for i in range(n - 1, 0, -1):
    line = ""
    for j in range(n - i):
        line += " "
    for k in range(2 * i - 1):
        line += "*"
    print(line)

#----------------------------------
# BETTER APPROACH (String multiplication)
#----------------------------------
n = 5
# Upper half
for i in range(1, n + 1):
    print(" " * (n - i) + "*" * (2 * i - 1))
# Lower half
for i in range(n - 1, 0, -1):
    print(" " * (n - i) + "*" * (2 * i - 1))

#----------------------------------
# BEST APPROACH (List comprehension + join)
#----------------------------------
n = 5
# Build both halves in one expression
pattern = [" " * (n - i) + "*" * (2 * i - 1) for i in range(1, n + 1)]
pattern += [" " * (n - i) + "*" * (2 * i - 1) for i in range(n - 1, 0, -1)]
print('\n'.join(pattern))

File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_09.cpp
//     *
//    ***
//   *****
//  *******
// *********
//  *******
//   *****
//    ***
//     *

//----------------------------------
// BRUTE FORCE APPROACH (Nested loops)
//----------------------------------
#include <iostream>
using namespace std;
int main()
{
    int n = 5;
    // Upper pyramid
    for (int i = 1; i <= n; i++)
    {
        // Print spaces
        for (int j = 1; j <= n - i; j++)
            cout << " ";
        // Print stars
        for (int k = 1; k <= 2 * i - 1; k++)
            cout << "*";
        cout << endl;
    }
    // Lower pyramid
    for (int i = n - 1; i >= 1; i--)
    {
        // Print spaces
        for (int j = 1; j <= n - i; j++)
            cout << " ";
        // Print stars
        for (int k = 1; k <= 2 * i - 1; k++)
            cout << "*";
        cout << endl;
    }
    return 0;
}

//----------------------------------
// BETTER APPROACH (String constructors per row)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    // Upper half
    for (int i = 1; i <= n; i++)
    {
        cout << string(n - i, ' ') << string(2 * i - 1, '*') << endl;
    }
    // Lower half
    for (int i = n - 1; i >= 1; i--)
    {
        cout << string(n - i, ' ') << string(2 * i - 1, '*') << endl;
    }
    return 0;
}

//----------------------------------
// BEST APPROACH (Pre-allocated buffer reuse)
//----------------------------------
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;
int main()
{
    const int n = 5;
    const int max_len = 2 * n - 1;
    string line(max_len, ' ');

    // Upper half
    for (int i = 1; i <= n; i++)
    {
        int spaces = n - i;
        int stars = 2 * i - 1;
        fill(line.begin(), line.begin() + spaces, ' ');
        fill(line.begin() + spaces, line.begin() + spaces + stars, '*');
        cout << line.substr(0, spaces + stars) << '\n';
    }

    // Lower half
    for (int i = n - 1; i >= 1; i--)
    {
        int spaces = n - i;
        int stars = 2 * i - 1;
        fill(line.begin(), line.begin() + spaces, ' ');
        fill(line.begin() + spaces, line.begin() + spaces + stars, '*');
        cout << line.substr(0, spaces + stars) << '\n';
    }
    return 0;
}